### PR TITLE
:seedling: Introduce re-pivoting e2e test

### PR DESCRIPTION
**Pull request by Rozzii**
This PR introduces the re-pivoting e2e test functionality.

### CODE REVIEW

**Comment by Rozzii** [comment]
/ test-v1a5-centos-e2e

File: test/e2e/remediation_test.go [hunk 1]
```diff
@@ -73,8 +73,8 @@ func test_remediation() {
 	By("Checking that rebooted node becomes Ready")
 	Logf("Marking a BMH '%s' for reboot", workerBmh.GetName())
 	annotateBmh(ctx, bootstrapClient, workerBmh, rebootAnnotation, pointer.String(""))
-	waitForVmsState([]string{vmName}, shutoff, specName)
-	waitForNodeStatus(ctx, targetClient, client.ObjectKey{Namespace: defaultNamespace, Name: workerNodeName}, v1.ConditionUnknown, specName)
+	//waitForVmsState([]string{vmName}, shutoff, specName)
```
Comment (Rozzii) [inline test/e2e/remediation_test.go:76]: I will remove these shortly.

File: test/e2e/node_reuse_test.go [hunk 2]
```diff
@@ -474,6 +476,12 @@ func node_reuse() {
 		e2eConfig.GetIntervals(specName, "wait-machine-running")...,
 	).Should(Equal(numberOfAllBmh))
 
+	By("Update Metal3MachineTemplate nodeReuse field to 'False'")
+	m3machineTemplateName := fmt.Sprintf("%s-controlplane", clusterName)
+	updateNodeReuse(true, m3machineTemplateName, targetClusterClient)
```
Comment (mboukhalfa) [inline test/e2e/node_reuse_test.go:481]:
```suggestion
	updateNodeReuse(false, m3machineTemplateName, targetClusterClient)
```

Also you need to edit updateNodeReuse function since it was always verifying that the value is true:
```
func updateNodeReuse(nodeReuse bool, m3machineTemplateName string, clusterClient client.Client) {
	m3machineTemplate := capm3.Metal3MachineTemplate{}
	Expect(clusterClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3machineTemplateName}, &m3machineTemplate)).To(Succeed())
	helper, err := patch.NewHelper(&m3machineTemplate, clusterClient)
	Expect(err).NotTo(HaveOccurred())
	m3machineTemplate.Spec.NodeReuse = nodeReuse
	Expect(helper.Patch(ctx, &m3machineTemplate)).To(Succeed())
	// verify that nodereuse is updated
	Expect(clusterClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3machineTemplateName}, &m3machineTemplate)).To(Succeed())
	Expect(m3machineTemplate.Spec.NodeReuse).To(Equal(nodeReuse))
}
```

File: test/e2e/node_reuse_test.go [hunk 3]
```diff
@@ -474,6 +476,12 @@ func node_reuse() {
...
+	m3machineTemplateName := fmt.Sprintf("%s-workers", clusterName)
+	updateNodeReuse(true, m3machineTemplateName, targetClusterClient)
```
Comment (mboukhalfa) [inline test/e2e/node_reuse_test.go:483]:
```suggestion
	updateNodeReuse(false, m3machineTemplateName, targetClusterClient)
```

Comment (Rozzii) [reply test/e2e/node_reuse_test.go:483]: Oh sorry,

Comment (Rozzii) [reply test/e2e/node_reuse_test.go]: Done

File: test/e2e/node_reuse_test.go (hunk 2 above)
Comment (Rozzii) [reply test/e2e/node_reuse_test.go]: Done

**Comment by Rozzii** [comment]
/ test-v1b1-centos-e2e

File: test/e2e/node_reuse_test.go [hunk 4]
```diff
@@ -473,6 +473,13 @@ func node_reuse() {
 		e2eConfig.GetIntervals(specName, "wait-machine-running")...,
 	).Should(Equal(numberOfAllBmh))
 
+	By("Update Metal3MachineTemplate nodeReuse field to 'False'")
+	m3machineTemplateName = fmt.Sprintf("%s-controlplane", clusterName)
+	updateNodeReuse(false, m3machineTemplateName, targetClusterClient)
+	m3machineTemplateName = fmt.Sprintf("%s-controlplane", clusterName)
+	updateNodeReuse(false, m3machineTemplateName, targetClusterClient)
```
Comment (furkatgofurov7) [inline test/e2e/node_reuse_test.go:480]: even though this should not be needed anymore, this is setting node reuse to false in the same m3mt

File: test/e2e/node_reuse_test.go [hunk 5]
```diff
@@ -509,7 +516,7 @@ func updateNodeReuse(nodeReuse bool, m3machineTemplateName string, clusterClient
 
 	// verify that nodereuse is true
 	Expect(clusterClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: m3machineTemplateName}, &m3machineTemplate)).To(Succeed())
-	Expect(m3machineTemplate.Spec.NodeReuse).To(BeTrue())
+	Expect(m3machineTemplate.Spec.NodeReuse).To(BeEquivalentTo(nodeReuse))
```
Comment (mboukhalfa) [inline test/e2e/node_reuse_test.go:519]: what's the difference here, I guess `BeTrue` more relevant ?

File: test/e2e/node_reuse_test.go (hunk 4 above)
Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: we discussed with mohamed that this PR would go in like this then when he refactors the node_reuse this will be removed in that PR.

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:480]: That refactoring will covered in #356

File: test/e2e/node_reuse_test.go (hunk 5 above)
Comment (Rozzii) [reply test/e2e/node_reuse_test.go:519]: This function is also used in case when we set the value to false not just when we set it to true that is why this has been modified.

File: test/e2e/node_reuse_test.go (hunk 4 above)
Comment (furkatgofurov7) [reply test/e2e/node_reuse_test.go:480]: @Rozzii and what happens if you do not set this to False in this PR? it should pass anyways because current implementation of e2e node reuse tests is not covering image upgrade part and following the old test workflow(where there is no new m3mt creation is introduced)

Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: It fails if it is not set to false, let's discuss this offline because I want to reflect on https://github.com/metal3-io/metal3-dev-env/pull/855 too but I haven't tested it yet. But anyways if it is not set to false it fails or at least it has failed until now, I will run a local test again just to make sure.

File: test/e2e/pivoting_test.go [hunk 6]
```diff
@@ -281,3 +290,113 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+
+	By("Check if BMH is in provisioned state")
+	Eventually(func() error {
+		bmhList := &bmo.BareMetalHostList{}
+		if err := bootstrapClusterProxy.GetClient().List(ctx, bmhList, client.InNamespace(namespace)); err != nil {
+			return err
+		}
+		for _, bmh := range bmhList.Items {
+			//if !(bmh.WasProvisioned() && bmh.Spec.Online) {
+			if !bmh.Spec.Online {
```
Comment (mboukhalfa) [inline test/e2e/pivoting_test.go:365]: Does online means that the bmh is provisioned

File: test/e2e/pivoting_test.go [hunk 7]
```diff
@@ -281,3 +290,113 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		Lister:      bootstrapClusterProxy.GetClient(),
+		ClusterName: pivotingCluster.Name,
+		Namespace:   pivotingCluster.Namespace,
+	})
+	Expect(controlPlane).ToNot(BeNil())
+
+	By("Check if BMH is in provisioned state")
+	Eventually(func() error {
+		bmhList := &bmo.BareMetalHostList{}
+		if err := bootstrapClusterProxy.GetClient().List(ctx, bmhList, client.InNamespace(namespace)); err != nil {
```
Comment (mboukhalfa) [inline test/e2e/pivoting_test.go:360]: if we can use expect to simplify the code

File: test/e2e/node_reuse_test.go (hunk 4 above)
Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: It is although true that there is a duplication by accident I will remove that.

Comment (mboukhalfa) [reply test/e2e/node_reuse_test.go:480]: That might means the issue raise only when CP is true ?

Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: Could be, we will now more after the local test has finished, I don't want to stress the CI at the moment.

Comment (furkatgofurov7) [reply test/e2e/node_reuse_test.go:480]: > That might mean the issue raise even if only the CP is true ?

I bet no, this should be removed completely and tests pass if we are following the exact same test flow as in m3-dev-env or previous workflow of it which does not include image upgrade (current node reuse e2e tests follow latter).

Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: It looks like the issue fully disappeared, I have removed this and as @furkatgofurov7  mentioned it has passed, but then i don't get it why wasn't it passing on v1alpha5.

Comment (Rozzii) [reply test/e2e/node_reuse_test.go:480]: So just to summarize the findings:

So this part of code that is highlighted above was locally (and partially in CI when it has worked) tested with both setting the CP and the worker nodes "node reuse" value to false and that was the only thing that made the test pass.

In the current PR i have just accidentally duplicated the CP node updating logic and deleted the worker node related part so this wasn't intentional!

Now in v1beta1 if  the "node reuse" value is not to false at the end of the nodeReuse test the test still passes, we have tested it on at least 4 machine both centos and ubuntu and in v1alpha5 it didn't work and now it does.

File: test/e2e/pivoting_test.go [hunk 8]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		By("Install Ironic in the target cluster")
+		installIronicBMO(bootstrapClusterProxy, "true", "false")
+
+		By("Reinstate Ironic Configmap")
+		configureIronicConfigmap(false)
+	}
+
+	By("Ensure API servers are stable before doing move")
+	// Nb. This check was introduced to prevent doing move to self-hosted in an aggressive way and thus avoid flakes.
+	// More specifically, we were observing the test failing to get objects from the API server during move, so we
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:324]:
```suggestion
	// More specifically, we observed the test failing to get objects from the API server during move, so we
```

File: test/e2e/pivoting_test.go [hunk 9]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	// are now testing the API servers are stable before starting move.
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:325]:
```suggestion
	// are now testing the API server to make sure it is stable before starting move.
```

File: test/e2e/pivoting_test.go [hunk 10]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	Consistently(func() error {
+		kubeSystem := &corev1.Namespace{}
+		return bootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKey{Name: "kube-system"}, kubeSystem)
+	}, "5s", "100ms").Should(BeNil(), "Failed to assert bootstrap API server stability")
+
+	By("Moving back to bootstrap cluster")
+	clusterctl.Move(ctx, clusterctl.MoveInput{
+		LogFolder:            filepath.Join(artifactFolder, "clusters", clusterName+"-bootstrap"),
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:333]:
```suggestion
		LogFolder:            filepath.Join(artifactFolder, "clusters", clusterName+"-pivot"),
```

File: test/e2e/pivoting_test.go [hunk 11]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		ClusterctlConfigPath: clusterctlConfigPath,
+		FromKubeconfigPath:   targetCluster.GetKubeconfigPath(),
+		ToKubeconfigPath:     bootstrapClusterProxy.GetKubeconfigPath(),
+		Namespace:            namespace,
+	})
+
+	LogFromFile(filepath.Join(artifactFolder, "clusters", clusterName+"-bootstrap", "clusterctl-move.log"))
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:340]:
```suggestion
	LogFromFile(filepath.Join(artifactFolder, "clusters", clusterName+"-pivot", "clusterctl-move.log"))
```

File: test/e2e/pivoting_test.go [hunk 12]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+
+	By("Checking that the repivoted cluster is up and running")
+	pivotingCluster := framework.DiscoveryAndWaitForCluster(ctx, framework.DiscoveryAndWaitForClusterInput{
+		Getter:    bootstrapClusterProxy.GetClient(),
+		Namespace: namespace,
+		Name:      clusterName,
+	}, e2eConfig.GetIntervals(specName, "wait-cluster")...)
+
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:348]: Should we have an Expect assertion here to check if pivotingCluster is not nil ?

File: test/e2e/pivoting_test.go [hunk 13]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		if err := bootstrapClusterProxy.GetClient().List(ctx, m3Machines, client.InNamespace(namespace)); err != nil {
+			return err
+		}
+		for _, m3Machine := range m3Machines.Items {
+			if !m3Machine.Status.Ready {
+				return errors.New("Metal3Machines cannot be ready")
+			}
+		}
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-object-provisioned")...).Should(BeNil())
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:379]:
```suggestion
	}, e2eConfig.GetIntervals(specName, "wait-object-provisioned")...).Should(Succeed())
```

File: test/e2e/pivoting_test.go [hunk 14]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		if err := bootstrapClusterProxy.GetClient().List(ctx, machines, client.InNamespace(namespace)); err != nil {
+			return err
+		}
+		for _, machine := range machines.Items {
+			if !strings.EqualFold(machine.Status.Phase, "running") { // Case insensitive comparison
+				return errors.New("Machines cannot be in the Running state")
+			}
+		}
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-machine-running")...).Should(BeNil())
```
Comment (namnx228) [inline test/e2e/pivoting_test.go:393]: Same here

**Review by namnx228 — CHANGES_REQUESTED** [review]
Great jobs. I left some comments below.

**Comment by Rozzii** [comment]
test-v1b1-integration

File: test/e2e/pivoting_test.go (hunk 14 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:393]: fixed

File: test/e2e/pivoting_test.go (hunk 13 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:379]: fixed.

File: test/e2e/pivoting_test.go (hunk 12 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:348]: I don't think it is needed, this DiscoveryAndWaitForCluster is full of checks and it can do waiting by itself too so I don't think we need a wrapper.

File: test/e2e/pivoting_test.go (hunk 11 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:340]: fixed

File: test/e2e/pivoting_test.go (hunk 10 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:333]: fixed

File: test/e2e/pivoting_test.go (hunk 9 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:325]: fixed

File: test/e2e/pivoting_test.go (hunk 8 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:324]: fixed

File: test/e2e/pivoting_test.go [hunk 15]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	if ephemeralCluster == "kind" {
+		bmoPath := os.Getenv("BMOPATH")
+		ironicCommand := bmoPath + "/tools/run_local_ironic.sh"
+		cmd := exec.Command("sh", "-c", "export CONTAINER_RUNTIME=docker; "+ironicCommand)
+		stdoutStderr, err := cmd.CombinedOutput()
+		fmt.Printf("%s\n", stdoutStderr)
+		Expect(err).To(BeNil(), "Cannot run local ironic")
+	} else {
+
+		LogFromFile(filepath.Join(artifactFolder, "clusters", clusterName+"-pivoting", "clusterctl-init.log"))
```
[the remark this answers is not in the document]
Comment (Rozzii) [reply test/e2e/pivoting_test.go:310]: fixed

File: test/e2e/pivoting_test.go [hunk 16]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	ephemeralCluster := os.Getenv("EPHEMERAL_CLUSTER")
+	if ephemeralCluster == "kind" {
+		bmoPath := os.Getenv("BMOPATH")
+		ironicCommand := bmoPath + "/tools/run_local_ironic.sh"
+		cmd := exec.Command("sh", "-c", "export CONTAINER_RUNTIME=docker; "+ironicCommand)
+		stdoutStderr, err := cmd.CombinedOutput()
+		fmt.Printf("%s\n", stdoutStderr)
+		Expect(err).To(BeNil(), "Cannot run local ironic")
+	} else {
+
```
[the remark this answers is not in the document]
Comment (Rozzii) [reply test/e2e/pivoting_test.go:309]: fixed

File: test/e2e/pivoting_test.go [hunk 17]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
 		Expect(err).To(BeNil(), "Cannot label BMO CRDs")
 	}
 }
+
+func rePivoting() {
+
+	By("By removing Ironic deployment from target cluster")
+	removeIronicDeploymentOnTarget()
+
+	By("Reinstate Ironic containers and BMHD")
```
[the remark this answers is not in the document]
Comment (Rozzii) [reply test/e2e/pivoting_test.go:299]: fixed,

File: test/e2e/pivoting_test.go [hunk 18]
```diff
@@ -281,3 +290,108 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
 		Expect(err).To(BeNil(), "Cannot label BMO CRDs")
 	}
 }
+
+func rePivoting() {
+
```
[the remark this answers is not in the document]
Comment (Rozzii) [reply test/e2e/pivoting_test.go:295]: fixed,

File: test/e2e/e2e_test.go [hunk 19]
```diff
@@ -81,10 +81,11 @@ var _ = Describe("Workload cluster creation", func() {
 			}, result)
 			cluster = result.Cluster
 			targetCluster = bootstrapClusterProxy.GetWorkloadCluster(ctx, namespace, clusterName)
-			test_remediation()
+			reMediation()
```
[the remark this answers is not in the document]
Comment (Rozzii) [reply test/e2e/e2e_test.go]: fixed

**Review by furkatgofurov7 — APPROVED** [review]
/approve

**Comment by fmuyassarov** [comment]
@Rozzii Can you please rebase with main because the issue must be resolved in the main branch.

File: test/e2e/cert_rotation_test.go [hunk 20]
```diff
@@ -19,8 +19,8 @@ import (
 	framework "sigs.k8s.io/cluster-api/test/framework"
 )
 
-func cert_rotation() {
-	Logf("Starting certificate rotation tests")
```
Comment (furkatgofurov7) [inline test/e2e/cert_rotation_test.go:23]: why is this removed?

File: test/e2e/node_reuse_test.go [hunk 21]
```diff
@@ -23,8 +23,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func node_reuse() {
-	Logf("Starting node reuse tests")
```
Comment (furkatgofurov7) [inline test/e2e/node_reuse_test.go:27]: why is this removed?

File: test/e2e/pivoting_test.go [hunk 22]
```diff
@@ -253,13 +254,21 @@ func removeIronicContainers() {
...
 	ironicNamespace := os.Getenv("IRONIC_NAMESPACE")
 	err := bootstrapClusterProxy.GetClientSet().AppsV1().Deployments(ironicNamespace).Delete(ctx, deploymentName, metav1.DeleteOptions{})
 	Expect(err).To(BeNil(), "Failed to delete Ironic from the source cluster")
 }
 
+func removeIronicDeploymentOnTarget() {
+	deploymentName := os.Getenv("NAMEPREFIX") + "-ironic"
+	ironicNamespace := os.Getenv("IRONIC_NAMESPACE")
+	err := targetCluster.GetClientSet().AppsV1().Deployments(ironicNamespace).Delete(ctx, deploymentName, metav1.DeleteOptions{})
+	Expect(err).To(BeNil(), "Failed to delete Ironic from the source cluster")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:269]: *target

File: test/e2e/pivoting_test.go [hunk 23]
```diff
@@ -282,3 +291,105 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
 		Expect(err).To(BeNil(), "Cannot label BMO CRDs")
 	}
 }
+
+func rePivoting() {
+	By("By removing Ironic deployment from target cluster")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:296]:
```suggestion
	By("Remove Ironic deployment from target cluster")
```

File: test/e2e/remediation_test.go [hunk 24]
```diff
@@ -39,8 +39,7 @@ const (
 
 const defaultNamespace = "default"
 
-func test_remediation() {
-	Logf("Starting remediation tests")
```
Comment (furkatgofurov7) [inline test/e2e/remediation_test.go:43]: why is this removed?

File: test/e2e/pivoting_test.go [hunk 25]
```diff
@@ -282,3 +291,105 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	By("Ensure API servers are stable before doing move")
+	// Nb. This check was introduced to prevent doing move to self-hosted in an aggressive way and thus avoid flakes.
+	// More specifically, it was observed that the test was failing to get objects from the API server during move, so now
+	// it is tested whether the API servers are stable before starting move.
+	Consistently(func() error {
+		kubeSystem := &corev1.Namespace{}
+		return bootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKey{Name: "kube-system"}, kubeSystem)
+	}, "5s", "100ms").Should(BeNil(), "Failed to assert bootstrap API server stability")
+
+	By("Moving back to bootstrap cluster")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:329]:
```suggestion
	By("Move back to bootstrap cluster")
```

File: test/e2e/pivoting_test.go [hunk 26]
```diff
@@ -282,3 +291,105 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+		LogFolder:            filepath.Join(artifactFolder, "clusters", clusterName+"-pivot"),
+		ClusterctlConfigPath: clusterctlConfigPath,
+		FromKubeconfigPath:   targetCluster.GetKubeconfigPath(),
+		ToKubeconfigPath:     bootstrapClusterProxy.GetKubeconfigPath(),
+		Namespace:            namespace,
+	})
+
+	LogFromFile(filepath.Join(artifactFolder, "clusters", clusterName+"-pivot", "clusterctl-move.log"))
+
+	By("Checking that the repivoted cluster is up and running")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:340]:
```suggestion
	By("Check that the re-pivoted cluster is up and running")
```

File: test/e2e/pivoting_test.go [hunk 27]
```diff
@@ -282,3 +291,105 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
...
+	pivotingCluster := framework.DiscoveryAndWaitForCluster(ctx, framework.DiscoveryAndWaitForClusterInput{
+		Getter:    bootstrapClusterProxy.GetClient(),
+		Namespace: namespace,
+		Name:      clusterName,
+	}, e2eConfig.GetIntervals(specName, "wait-cluster")...)
+
+	By("Checking that the control plane of the repivoted cluster is up and running")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:347]:
```suggestion
	By("Check that the control plane of the re-pivoted cluster is up and running")
```

**Review by furkatgofurov7 — CHANGES_REQUESTED** [review]
/approve cancel

File: test/e2e/cert_rotation_test.go (hunk 20 above)
Comment (Rozzii) [reply test/e2e/cert_rotation_test.go:23]: wasn't removed just renamed.

File: test/e2e/pivoting_test.go (hunk 23 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:296]: fixed

File: test/e2e/pivoting_test.go (hunk 22 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:269]: fixed

File: test/e2e/node_reuse_test.go (hunk 21 above)
Comment (Rozzii) [reply test/e2e/node_reuse_test.go:27]: fixed

File: test/e2e/pivoting_test.go (hunk 25 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:329]: fixed

File: test/e2e/pivoting_test.go (hunk 26 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:340]: fixed

File: test/e2e/pivoting_test.go (hunk 27 above)
Comment (Rozzii) [reply test/e2e/pivoting_test.go:347]: fixed

File: test/e2e/remediation_test.go (hunk 24 above)
Comment (Rozzii) [reply test/e2e/remediation_test.go:43]: fixed

File: test/e2e/pivoting_test.go [hunk 28]
```diff
@@ -282,3 +291,105 @@ func labelBMOCRDs(targetCluster framework.ClusterProxy) {
 		Expect(err).To(BeNil(), "Cannot label BMO CRDs")
 	}
 }
+
+func rePivoting() {
+	By("Remove Ironic deployment from target cluster")
```
Comment (furkatgofurov7) [inline test/e2e/pivoting_test.go:296]: Let's add a log for beginning of re-pivoting tests too?

Comment (Rozzii) [reply test/e2e/pivoting_test.go:296]: okay

**Review by furkatgofurov7 — APPROVED** [review]
Thanks @Rozzii for addressing all comments, let's merge it this time :grinning:

/approve
/cc @namnx228

File: test/e2e/config/e2e_conf.yaml [hunk 29]
```diff
@@ -89,7 +89,7 @@ intervals:
   default/wait-delete-cluster: ["20m", "10s"]
   default/wait-machine-upgrade: ["50m", "10s"]
   default/wait-machine-remediation: ["30m", "10s"]
-  default/wait-vm-state: ["10m", "100ms"]
+  default/wait-vm-state: ["20m", "100ms"]
```
Comment (namnx228) [inline test/e2e/config/e2e_conf.yaml:92]: If I am not wrong, this change is for the bug in remediation reported by community, right? It is fixed in https://github.com/metal3-io/cluster-api-provider-metal3/pull/420. So maybe we can keep the time to 10m

**Review by namnx228 — COMMENTED** [review]
Look good to me, just a small nit, then this PR is good to go.

File: test/e2e/config/e2e_conf.yaml (hunk 29 above)
Comment (Rozzii) [reply test/e2e/config/e2e_conf.yaml:92]: It was partly but I left it there because I have seen it needing a bit more time when I was testing on laptop.
